internal/typeinfo/typeinfo.go

```diff
@@ -0,0 +1,115 @@
+package typeinfo
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+var cacheMutex sync.RWMutex
+var cache = make(map[reflect.Type]*Info)
+
+// Reflect will return the Info of a given type,
+// generating and caching as required.
+func TypeInfo(value any) (*Info, error) {
+	if value == (any)(nil) {
+		return &Info{}, fmt.Errorf("cannot reflect nil value")
+	}
+
+	v := reflect.ValueOf(value)
+	v = reflect.Indirect(v)
+
+	cacheMutex.RLock()
+	info, found := cache[v.Type()]
+	cacheMutex.RUnlock()
+	if found {
+		return info, nil
+	}
+
+	info, err := generate(v)
+	if err != nil {
+		return &Info{}, err
+	}
+
+	cacheMutex.Lock()
+	cache[v.Type()] = info
+	cacheMutex.Unlock()
+
+	return info, nil
+}
+
+// generate produces and returns reflection information for the input
+// reflect.Value that is specifically required for SQLAir operation.
+func generate(value reflect.Value) (*Info, error) {
+	// Dereference the value if it is a pointer.
+	value = reflect.Indirect(value)
+
+	// Reflection information is only generated for structs.
+	if value.Kind() != reflect.Struct {
+		return &Info{}, fmt.Errorf("cannot reflect type %q, only struct", value.Kind())
```

Comment:
Same note about zero values on error above, and the error message is also not very clear, as we certainly can reflect on any type we want in Go.  This error should say something like: "internal error: attempted to obtain struct information for something that is not a struct: %T.   Also, the parameter is a Kind, not a Type, and it would indeed be more useful to have the type name here so that it's easier for the developer to find the offending code.

Reply:
Using type name now.